File: integrations/defold/wyrdforge/lua/wyrdforge.lua
--- wyrdforge.lua — WyrdForge HTTP client module for Defold (Phase 11F).
--
-- Provides the full WYRD Protocol Lua API for Defold games.
-- Uses Defold's built-in http and json modules for async HTTP.
-- Requires the wyrdforge native extension (src/wyrdforge.cpp) to be loaded.
--
-- SETUP
--   1. Add the wyrdforge extension to your game.project dependencies.
--   2. Require this module in any script that needs WYRD.
--
-- QUICK START
--   local wyrd = require("wyrdforge.lua.wyrdforge")
--
--   function init(self)
--       wyrd.init("localhost", 8765)
--   end
--
--   function on_input(self, action_id, action)
--       if action_id == hash("talk") and action.pressed then
--           wyrd.query("sigrid", "What do the runes say?", function(ok, response, err)
--               if ok then
--                   msg.post("/npc#dialogue", "show_text", {text = response})
--               else
--                   print("[WyrdForge] Query error:", err)
--               end
--           end)
--       end
--   end
--
-- API REFERENCE
--   wyrd.init(host, port)
--   wyrd.set_enabled(bool)
--   wyrd.query(persona_id, user_input, callback)
--   wyrd.push_observation(title, summary, callback)
--   wyrd.push_fact(subject_id, key, value, callback)
--   wyrd.health(callback)

local M = {}

-- ---------------------------------------------------------------------------
-- Internal state
-- ---------------------------------------------------------------------------

local _host    = "localhost"
local _port    = 8765
local _enabled = true

-- ---------------------------------------------------------------------------
-- Private helpers
-- ---------------------------------------------------------------------------

local function _base_url()
    return string.format("http://%s:%d", _host, _port)
end

--- Fire a POST request to WyrdHTTPServer.
-- @param path        string     e.g. "/query" or "/event"
-- @param body_json   string     pre-built JSON body string
-- @param on_raw      function   called with (ok, decoded_table_or_nil, err_string_or_nil)
local function _post(path, body_json, on_raw)
    if not _enabled then return end
    local headers = {["Content-Type"] = "application/json"}
    http.request(
        _base_url() .. path,
        "POST",
        function(self, id, response)
            if response.status >= 200 and response.status < 300 then
                local ok, data = pcall(json.decode, response.response)
                if ok and data ~= nil then
                    if on_raw then on_raw(true, data, nil) end
                else
                    if on_raw then on_raw(false, nil, "WyrdForge: invalid JSON response") end
                end
            else
                local msg = string.format("WyrdForge: HTTP %d", response.status)
                if on_raw then on_raw(false, nil, msg) end
            end
        end,
        headers,
        body_json
    )
end

-- ---------------------------------------------------------------------------
-- Public API
-- ---------------------------------------------------------------------------

--- Initialize WyrdForge. Call once from your game script's init() function.
-- @param host  string   WyrdHTTPServer hostname (default: "localhost")
-- @param port  number   WyrdHTTPServer port (default: 8765)
function M.init(host, port)
    _host    = host or "localhost"
    _port    = port or 8765
    _enabled = true
end

--- Enable or disable WyrdForge without changing host/port.
-- @param enabled boolean
function M.set_enabled(enabled)
    _enabled = enabled == true
end

--- Query WYRD world context for a character.
-- Fires an async HTTP POST to /query. The callback runs when the response arrives.
-- @param persona_id  string    WYRD persona ID (e.g. "sigrid")
-- @param user_input  string    Player message or query text. Empty string = default context.
-- @param callback    function  function(ok, response_text, error_msg)
--                              ok            boolean — true on success
--                              response_text string  — the character context string
--                              error_msg     string  — error message on failure, nil on success
function M.query(persona_id, user_input, callback)
    local body = wyrdforge.build_query_body(persona_id or "", user_input or "")
    _post("/query", body, function(ok, data, err)
        if ok then
            if callback then callback(true, data.response or "", nil) end
        else
            if callback then callback(false, nil, err) end
        end
    end)
end

--- Push a world event observation to WYRD memory.
-- Fires an async HTTP POST to /event. Use to record things that happened in the game world.
-- @param title     string    Short title for the observation (e.g. "Dragon appears")
-- @param summary   string    Description of what happened
-- @param callback  function  function(ok, data, error_msg) — optional
function M.push_observation(title, summary, callback)
    local payload = json.encode({title = title or "", summary = summary or ""})
    local body = wyrdforge.build_event_body("observation", payload)
    _post("/event", body, callback)
end

--- Push a canonical world fact about an entity to WYRD.
-- Fires an async HTTP POST to /event. Use to record persistent facts.
-- @param subject_id  string    Entity or character ID this fact is about
-- @param key         string    Fact key (e.g. "location", "role", "alignment")
-- @param value       string    Fact value
-- @param callback    function  function(ok, data, error_msg) — optional
function M.push_fact(subject_id, key, value, callback)
    local payload = json.encode({
        subject_id = subject_id or "",
        key        = key or "",
        value      = value or "",
    })
    local body = wyrdforge.build_event_body("fact", payload)
    _post("/event", body, callback)
end

--- Check whether WyrdHTTPServer is reachable and healthy.
-- @param callback  function  function(ok, data, error_msg)
--                            ok   boolean — true if server responds with status="ok"
function M.health(callback)
    if not _enabled then return end
    http.request(
        _base_url() .. "/health",
        "GET",
        function(self, id, response)
            if response.status == 200 then
                local ok, data = pcall(json.decode, response.response)
                if ok and data ~= nil then
                    if callback then callback(data.status == "ok", data, nil) end
                else
                    if callback then callback(false, nil, "WyrdForge: invalid JSON on health check") end
                end
            else
                local msg = string.format("WyrdForge: health check HTTP %d", response.status)
                if callback then callback(false, nil, msg) end
            end
        end,
        {},
        nil
    )
end

return M

File: integrations/roll20/wyrdforge/wyrdforge.js
/**
 * WyrdForge — WYRD Protocol integration for Roll20 API (Phase 10B).
 *
 * Roll20 API script. Requires Roll20 Pro (API access).
 * Communicates with WyrdHTTPServer via Roll20's `sendChat` + external fetch
 * (Roll20 API scripts run in Node.js and have access to require/https).
 *
 * Features:
 *   !wyrd <persona_id> [query]   — chat command to query world context
 *   !wyrd-sync <actor_name>      — register a character as a WYRD entity
 *   !wyrd-health                 — check server connectivity
 *
 * Setup:
 *   1. In Roll20 API console, set the following state defaults (or edit below):
 *      state.wyrdforge = { host: "your-server", port: 8765, enabled: true }
 *   2. Add this script to your Roll20 campaign API scripts.
 *   3. Run !wyrd-health to verify connectivity.
 */

/* global on, sendChat, state, log, getAttrByName */

const WF_MODULE = "WyrdForge";
const WF_CMD = "!wyrd";
const WF_SYNC_CMD = "!wyrd-sync";
const WF_HEALTH_CMD = "!wyrd-health";

// ---------------------------------------------------------------------------
// Config helpers (pure, exportable for tests)
// ---------------------------------------------------------------------------

/**
 * @param {object} stateObj  — Roll20 state object
 * @returns {{ host: string, port: number, enabled: boolean }}
 */
export function getConfig(stateObj) {
  if (!stateObj.wyrdforge) {
    stateObj.wyrdforge = { host: "localhost", port: 8765, enabled: true };
  }
  return stateObj.wyrdforge;
}

// ---------------------------------------------------------------------------
// Chat command parser (pure, exportable for tests)
// ---------------------------------------------------------------------------

/**
 * Parse a Roll20 chat message content string.
 * @param {string} message
 * @returns {{ cmd: string|null, personaId: string, query: string }}
 */
export function parseCommand(message) {
  const trimmed = (message || "").trim();

  if (trimmed === WF_HEALTH_CMD) {
    return { cmd: "health", personaId: "", query: "" };
  }

  if (trimmed.startsWith(WF_SYNC_CMD)) {
    const rest = trimmed.slice(WF_SYNC_CMD.length).trim();
    return { cmd: "sync", personaId: rest, query: "" };
  }

  if (trimmed.startsWith(WF_CMD)) {
    const rest = trimmed.slice(WF_CMD.length).trim();
    const spaceIdx = rest.indexOf(" ");
    if (spaceIdx === -1) {
      return { cmd: "query", personaId: rest, query: "" };
    }
    return {
      cmd: "query",
      personaId: rest.slice(0, spaceIdx),
      query: rest.slice(spaceIdx + 1).trim(),
    };
  }

  return { cmd: null, personaId: "", query: "" };
}

// ---------------------------------------------------------------------------
// Persona ID normalizer (pure, exportable for tests)
// ---------------------------------------------------------------------------

/**
 * @param {string} name
 * @returns {string}
 */
export function normalizePersonaId(name) {
  return (name || "").toLowerCase().replace(/[^a-z0-9_]/g, "_").replace(/_+/g, "_").slice(0, 64);
}

// ---------------------------------------------------------------------------
// Chat message formatter (pure, exportable for tests)
// ---------------------------------------------------------------------------

/**
 * Format a WYRD response as a Roll20 chat output string.
 * @param {string} personaId
 * @param {string} responseText
 * @returns {string}
 */
export function formatChatOutput(personaId, responseText) {
  const divider = "—".repeat(40);
  return `/w gm &{template:default} {{name=ᚹ WyrdForge — ${personaId}}} {{${divider}=${responseText}}}`;
}

/**
 * Format a simple status/info message for GM whisper.
 * @param {string} message
 * @returns {string}
 */
export function formatStatus(message) {
  return `/w gm [WyrdForge] ${message}`;
}

// ---------------------------------------------------------------------------
// HTTP client helpers (pure logic, testable with mock http)
// ---------------------------------------------------------------------------

/**
 * Build the options for a GET request to WyrdHTTPServer.
 * @param {string} host
 * @param {number} port
 * @param {string} path
 * @returns {{ hostname: string, port: number, path: string, method: string }}
 */
export function buildGetOptions(host, port, path) {
  return { hostname: host, port, path, method: "GET" };
}

/**
 * Build the options and body for a POST request to WyrdHTTPServer.
 * @param {string} host
 * @param {number} port
 * @param {string} path
 * @param {object} payload
 * @returns {{ options: object, body: string }}
 */
export function buildPostOptions(host, port, path, payload) {
  const body = JSON.stringify(payload);
  return {
    options: {
      hostname: host,
      port,
      path,
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(body),
      },
    },
    body,
  };
}

// ---------------------------------------------------------------------------
// HTTP request wrapper (uses Node.js http module — Roll20 API environment)
// ---------------------------------------------------------------------------

/**
 * Make an HTTP request. Returns a Promise.
 * @param {object} http   — Node.js `http` module (or mock)
 * @param {object} opts   — http.request options
 * @param {string} [body] — POST body
 * @returns {Promise<object>}
 */
export function httpRequest(http, opts, body = null) {
  return new Promise((resolve, reject) => {
    const req = http.request(opts, (res) => {
      let raw = "";
      res.on("data", (chunk) => { raw += chunk; });
      res.on("end", () => {
        try {
          resolve({ status: res.statusCode, data: JSON.parse(raw) });
        } catch (err) {
          reject(new Error(`JSON parse error: ${err.message}`));
        }
      });
    });
    req.on("error", reject);
    if (body) req.write(body);
    req.end();
  });
}

// ---------------------------------------------------------------------------
// High-level operations (use http + config)
// ---------------------------------------------------------------------------

/**
 * Query WYRD world context for a persona.
 * @param {object} http
 * @param {{ host: string, port: number }} config
 * @param {string} personaId
 * @param {string} [query=""]
 * @returns {Promise<string>}  — response text
 */
export async function queryWyrd(http, config, personaId, query = "") {
  const payload = {
    persona_id: personaId,
    user_input: query || "What is the current state of the world?",
    use_turn_loop: false,
  };
  const { options, body } = buildPostOptions(config.host, config.port, "/query", payload);
  const result = await httpRequest(http, options, body);
  return result.data?.response ?? "";
}

/**
 * Push a fact to WYRD memory.
 * @param {object} http
 * @param {{ host: string, port: number }} config
 * @param {string} personaId
 * @param {string} name
 * @returns {Promise<boolean>}
 */
export async function syncActor(http, config, personaId, name) {
  const payload = {
    event_type: "fact",
    payload: { subject_id: personaId, key: "name", value: name },
  };
  const { options, body } = buildPostOptions(config.host, config.port, "/event", payload);
  const result = await httpRequest(http, options, body);
  return result.data?.ok === true;
}

/**
 * Health-check the WyrdHTTPServer.
 * @param {object} http
 * @param {{ host: string, port: number }} config
 * @returns {Promise<boolean>}
 */
export async function checkHealth(http, config) {
  const opts = buildGetOptions(config.host, config.port, "/health");
  const result = await httpRequest(http, opts);
  return result.data?.status === "ok";
}

// ---------------------------------------------------------------------------
// Roll20 entry point — only runs in Roll20 API environment
// ---------------------------------------------------------------------------

/* c8 ignore start */
if (typeof on !== "undefined") {
  const http = require("http"); // eslint-disable-line no-undef

  on("ready", () => {
    log(`${WF_MODULE}: loaded. Type !wyrd-health to check server connectivity.`);
  });

  on("chat:message", async (msg) => {
    if (msg.type !== "api") return;
    const cfg = getConfig(state);
    if (!cfg.enabled) return;

    const parsed = parseCommand(msg.content);
    if (!parsed.cmd) return;

    const speaker = msg.who || "API";

    try {
      if (parsed.cmd === "health") {
        const ok = await checkHealth(http, cfg);
        sendChat(WF_MODULE, formatStatus(ok ? "WYRD server is online." : "WYRD server unreachable."));

      } else if (parsed.cmd === "sync") {
        if (!parsed.personaId) {
          sendChat(WF_MODULE, formatStatus("Usage: !wyrd-sync <actor_name>"));
          return;
        }
        const pid = normalizePersonaId(parsed.personaId);
        const ok = await syncActor(http, cfg, pid, parsed.personaId);
        sendChat(WF_MODULE, formatStatus(ok ? `Synced '${pid}' to WYRD.` : `Sync failed for '${pid}'.`));

      } else if (parsed.cmd === "query") {
        if (!parsed.personaId) {
          sendChat(WF_MODULE, formatStatus("Usage: !wyrd <persona_id> [query]"));
          return;
        }
        const pid = normalizePersonaId(parsed.personaId);
        const response = await queryWyrd(http, cfg, pid, parsed.query);
        sendChat(WF_MODULE, formatChatOutput(pid, response));
      }
    } catch (err) {
      sendChat(WF_MODULE, formatStatus(`Error: ${err.message}`));
      log(`${WF_MODULE} error: ${err.message}`);
    }
  });
}
/* c8 ignore end */
